pkg/ovsdb/monitors.go

```diff
@@ -5,198 +5,262 @@ import (
-		len(m.insert[&m.updaters[indx]]) == 0 {
-		// we can remove this handler
-		m.updaters[indx] = m.updaters[len(m.updaters)-1]
-		m.updaters = m.updaters[:len(m.updaters)-1]
-		if len(m.updaters) == 0 {
-			m.cancel()
-			return true
-		}
-	}
-	return false
+	return
 }
 
-func newMonitor(cli *clientv3.Client, prefix string, mcr ovsjson.MonitorCondRequest, isV1 bool, handler *handlerKey) *monitor {
-
-	ctx := context.Background()
-	ctxt, cancel := context.WithCancel(ctx)
-	wch := cli.Watch(clientv3.WithRequireLeader(ctxt), prefix,
-		clientv3.WithPrefix(),
-		clientv3.WithCreatedNotify(),
-		clientv3.WithPrevKV())
-
-	monitor := &monitor{cancel: cancel}
+func (m *monitor) start() {
 	go func() {
-		for wresp := range wch {
-			for _, ev := range wresp.Events {
-				var dest map[*updater][]*handlerKey
-				monitor.mu.Lock()
-				if ev.IsCreate() {
-					// insert event
-					dest = monitor.insert
-				} else if ev.IsModify() {
-					dest = monitor.modify
-				} else {
-					// delete event
-					dest = monitor.delete
+		for wresp := range m.watchChannel {
+			if wresp.Canceled {
+				// TODO should we just reconnect
+				m.mu.Lock()
+				for hlk := range m.handlers {
+					// run in separate goroutines
+					go hlk.handler.monitorCanceledNotification(hlk.jsonValue)
 				}
-				monitor.mu.Unlock()
-				// TODO queue or a separate goroutine
-				monitor.propagateEvent(ev, dest)
+				m.mu.Unlock()
+				// remove itself
+				m.db.RemoveMonitor(m.dataBaseName)
+				return
+			}
+			result, _ := m.prepareTableUpdate(wresp.Events)
+			for hd, tu := range result {
+				go hd.handler.notify(hd.jsonValue, tu)
 			}
 		}
 	}()
-	return monitor
 }
 
-func (m *monitor) propagateEvent(event *clientv3.Event, destinations map[*updater][]*handlerKey) {
-	// for each updater entry create a return data and call each handler
+func (m *monitor) hasHandlers() bool {
+	return len(m.handlers) > 0
 }
 
-func (u *updater) prepareTableUpdate(event *clientv3.Event) (table string, uuid string, rowUpdate *ovsjson.RowUpdate, err error) {
-	key := string(event.Kv.Key)
-	if table, uuid, err = keyToTableAndUUID(key); err != nil {
-		return
+func mcrToUpdater(mcr ovsjson.MonitorCondRequest, isV1 bool) *updater {
+	sort.Strings(mcr.Columns)
+	var key string
+	for _, c := range mcr.Columns {
+		key = key + c
+	}
+	// TODO handle "Where"
+	if mcr.Select == nil {
+		mcr.Select = &allTrueSelect
+	}
+	key = fmt.Sprintf("%s%v%v", key, isV1, *mcr.Select)
+	return &updater{Columns: common.StringArrayToMap(mcr.Columns), isV1: isV1, Select: *mcr.Select, key: key}
+}
+
+func (m *monitor) prepareTableUpdate(events []*clientv3.Event) (map[handlerKey]ovsjson.TableUpdates, error) {
+	// prepare results that will be sent to clients
+	// handlerKey -> tableName -> uuid -> rowUpdate
+	result := map[handlerKey]ovsjson.TableUpdates{}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, ev := range events {
+		key := string(ev.Kv.Key)
+		path, table, uuid, err := keyToTableAndUUID(key)
```

Comment:
the path should be broken up into:
- server/service (e.g. ndb, sdb, etc)
- dbname (_Server, OVN_Southbound)